# 处理库代码中的错误

`Lua` 是一种**安全**（ `safe` ）的语言。这意味着不管用 `Lua` 写的是什么，也不管写出来的内容多么不正确，我们总是能用它自身的机制来理解程序的行为。此外，程序中的错误也是通过 `Lua` 语言的机制来检测和解释的。与之相比，许多 `C` 语言代码中的错误只能从底层硬件的角度来解释（例如，把异常位置作为指令地址给出）。

只要往 `Lua` 中加入新的 `C` 函数，这种安全性就可能被打破。例如，一个等价于 `BASIC` 命令 `poke` 的函数（该函数用于将任意的字节存储到任意的内存地址中）就可能导致各种各样的内存崩溃。因此，我们必须确保新加入的内容对 `Lua` 语言来说是安全的，并提供妥善的错误处理。

正如之前所讨论的， `C` 语言程序必须通过 `lua_pcall` 设置错误处理。不过，在为 `Lua` 编写库函数时，通常无需处理错误。库函数抛出的错误要么被 `Lua` 中的 `pcall` 捕获，要么被应用代码中的 `lua_pcall` 捕获。因此，当 `C` 语言库中的函数检测到错误时，只需要简单的调用 `lua_error` 即可（或调用 `luaL_error` 更好，它会格式化错误信息，然后调用 `lua_error` ）。函数 `lua_error` 会收拾 `Lua` 系统中的残局，然后跳转回保护模式调用处，并传递错误信息。
